### ISSUE TITLE
apps.json has SideStore compatibility issues since October 18

### ISSUE THREAD

**Issue by jgottula** [arichornlover/arichornlover.github.io#13]
I'll preface this by saying that this is really ultimately a problem on the SideStore side, involving the upstream AltStore source file format spec continually changing over time, and SideStore lagging behind somewhat on maintaining full compatibility with it.

But, some changes a few weeks ago here to the source file JSON for uYouEnhanced recently made it SideStore-incompatible. And the workaround to make it SideStore-compatible for the time being is simple enough that it may be worth just implementing that until they get a chance to get their AltStore source format compatibility situation a little more under control.

I filed an upstream bug report, which goes into detail on the particular incompatibility that's at issue here: https://github.com/SideStore/SideStore/issues/735

---

But the short version is this:

In 7bd125b, having added a "versions" array for uYouEnhanced with entries for 19.40.4 and 19.20.2, you removed the _top-level_ (non-version specific) "downloadURL" key (which was essentially just duplicative of the "downloadURL" key for the 19.40.4 version).

That would be fine. But SideStore, at least currently (as of the most recent release, 0.5.8, as well as the latest git at the moment), still expects the top-level "downloadURL" key to be there, even if there's a "versions" array; and if it's missing, then it considers that to be a fatal defect that makes the whole file invalid.

**So the workaround is to basically just do the opposite of what 7bd125b did: add back the top-level "downloadURL" key. (It should contain the same URL as the latest version's "downloadURL" key.)**

---

Hopefully this will get fixed in SideStore relatively soon. If I had an Xcode environment just sitting around, I would write up a PR myself right now: I can see exactly what's going on between the relevant parts of the AltStore and SideStore code, as well as what would need to be changed to fix it; but since I can't readily compile-check, nor runtime-test, any changes I'd make (plus I'm not especially familiar with Swift and so I'd probably goof up some things here and there), unfortunately I'm left in the position of just filing a detailed bug report with them and hoping that that's enough to get the ball rolling.

**Comment by arichornlover** [comment]
I did some updates, is it working for SideStore now?

**Comment by jgottula** [comment]
> I did some updates, is it working for SideStore now?

Yes! I think within a few days at most from when I first posted this issue, things were working properly again on SideStore 0.5.8. And I'm continuing to see everything working fine now using SideStore 0.5.9 as well.

---

**EDIT: See the comment following this one... things are in fact not actually fixed yet to where a workaround isn't needed.**

Incidentally, I'm not actually entirely sure that this workaround is necessarily needed for much longer, since [SideStore 0.5.9](https://github.com/SideStore/SideStore/releases/tag/0.5.9), released on Dec 24, had [some changes](https://github.com/SideStore/SideStore/pull/807) that I think were meant to address this exact issue. (I haven't done any specific testing of those changes, but they do seem to be literally for this specific situation involving versioned apps lacking a base downloadURL.)

Also of note, they released [0.5.10](https://github.com/SideStore/SideStore/releases/tag/0.5.10) a few days later, with unspecified changes related to "source issues". (It appears to work fine; no obvious issues.)

And then 0.6.0, which is coming sometime soon-ish, and is currently in alpha, I believe finally properly merges in a lot of AltStore 2.0 stuff, which should presumably bring the app source file format compatibility situation into a much better place overall. (I think prior to this they've been based on an older branch of AltStore, and so whenever AltStore 2.x made this-or-that change to the source file format, it'd just kinda make stuff break for SideStore until they came up with their own one-off band-aid compat fix for the breakage.)

**Comment by jgottula** [comment]
Huh... okay so I did some quick testing just now, with some other sources that I knew to have the issue with SideStore (where they do have a versions array but don't have a downloadURL outside the versions array).

And actually, it appears that SideStore 0.5.10 does not fix the problem. It still hates sources like that. It just gives a different error message now. Yay!
```
No value associated with key CodingKeys(string value: "version", intValue: nil) ("version").
```

So maybe 0.6.0 (or a 0.5.11 or whatever) will actually finally fix this dumb thing for real... but until then, the workaround is still necessary.

---

But anyway, just to be perfectly clear:
- the change done in 0338f95 (ditching the versions array entirely) worked totally fine
- the current revision 97387f5 (which has a versions array, but which also makes sure to have a downloadURL key outside the versions array too) also continues to work totally fine
